src/graphql/mirror.js

```diff
@@ -728,6 +740,68 @@ export class Mirror {
     }
   }
 
+  /**
+   * Save the request query and query parameters to the database.
+   */
+
+  async _log_request(
+    body: Queries.Body,
+    variables: {+[string]: any}
+  ): Promise<number> {
+    const query = Queries.stringify.body(body, Queries.inlineLayout());
+    const query_parameters = JSON.stringify(variables);
+    const request_timestamp = new Date();
+
+    const log = this._db.prepare(
+      dedent`\
+        INSERT INTO network_log (query, query_parameters, request_timestamp)
+        VALUES (?, ?, ?)
+      `
+    );
+
+    return await log.run(query, query_parameters, +request_timestamp)
+      .lastInsertRowid;
+  }
+
+  /**
+   * Save the network response and response timestamp to the table row
+   * corresponding to the request that generated it.
+   */
+
+  _log_response(rowid: number, jsonResponse: UpdateResult): void {
+    const response = JSON.stringify(jsonResponse);
+    const response_timestamp = new Date();
+    this._db
+      .prepare(
+        dedent`\
+        UPDATE network_log
+        SET
+            response = ?,
+            response_timestamp = ?
+        WHERE rowid = ?
```

Comment:
If the provided row ID does not actually match any rows, this code as
written will silently succeed. The `_makeSingleUpdateFunction` helper
can handle the relevant checks and errors for you; you should use it.

Reply:
updated